### PR TITLE
Eliminate Remaining `requests` usage within `azure-sdk-tools`

### LINKED ISSUE #26925: Remove tools dependencies on `requests`
We are headed towards a world where we can **optionally** install sync or async `requests` alternatives. To properly test presence and **non-presence**  of these dependencies, we need to ensure that our tooling doesn't use any of them.

This issue covers replacing all usage of the `requests` package internal to `tools/azure-sdk-tools` and `tools/azure-devtools` with `urllib3`.

Recommendations are welcomed, I just found that one as the most popular alternative.

### PR DESCRIPTION
**Pull request by scbedd**
Resolves #26925

Remaining to complete:

- [x] Raise for status when sanitizer, matcher, or transform add fails. We weren't seeing the sanitizers raise before because we weren't using the `sanitizers.py` http client. 
- [x] Ensure the new `urllib3` honors the overridden environment variables that we use to override requests cert bundles.

### CODE REVIEW

**Comment by scbedd** [comment]
> Should we also remove the `requests` imports in `proxy_startup.py` and `proxy_testcase.py`?

**Walter White Gif**

**Review by mccoyp — APPROVED** [review]

File: tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py [hunk 1]
```diff
@@ -58,36 +62,39 @@ def get_recording_assets(test_id: str) -> str:
...
+            url=PLAYBACK_START_URL,
+            fields=json_payload,
         )
-        if result.status_code != 200:
-            message = six.ensure_str(result._content)
+        if result.status != 200:
+            message = six.ensure_str(result.data)
             raise HttpResponseError(message=message)
 
         try:
```
Comment (pvaneck) [inline tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py:100]: Line 104: `result.text` - > `result.data`
Line 106: `variables = json.loads(result.data.decode('utf-8'))`

File: tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py [hunk 2]
```diff
@@ -58,36 +62,39 @@ def get_recording_assets(test_id: str) -> str:
...
+        json_payload["x-recording-assets-file"] = assets_json
 
     if is_live():
-        result = requests.post(
-            RECORDING_START_URL,
-            json=json_payload,
+        result = http_client.request(
+            method="POST",
+            url=RECORDING_START_URL,
+            fields=json_payload,
```
Comment (pvaneck) [inline tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py:83]: After testing, using `fields` doesn't seem to work. Json encoding and using `body` does work fine though:

i.e.:
```
json_payload = json.dumps(json_payload).encode('utf-8')
result = http_client.request(
    method="POST",
    url=RECORDING_START_URL,
    body=json_payload
)
```

I would replace all usage of `fields` in `http_client.request` to use `body`.

File: tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py [hunk 3]
```diff
@@ -109,25 +116,26 @@ def start_record_or_playback(test_id: str) -> "Tuple[str, Dict[str, str]]":
...
+            http_client.request(
+                method="POST",
+                url=RECORDING_STOP_URL,
+                headers={
+                    "x-recording-file": test_id,
+                    "x-recording-id": recording_id,
+                    "x-recording-save": "true",
+                    "Content-Type": "application/json",
+                },
+                json=test_variables or {},  # tests don't record successfully unless test_variables is a dictionary
```
Comment (pvaneck) [inline tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py:130]: use `body` instead of `json`

**Review by pvaneck — COMMENTED** [review]
Hey @scbedd. Ran into some issues with `requests` not being available in certain pipeline environments ([example](https://dev.azure.com/azure-sdk/public/_build/results?buildId=2037071&view=logs&j=4fa73ede-de59-50b0-851d-db96582ea108&t=271dca1a-6c62-5d1b-458d-2dd10be9b8db&l=10994)), and that led me here, which should resolve that issue. However, after running playback tests on a service using this PR (to test it), I ran into the following issue:

```
../../../tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:193: in start_test_proxy
    set_custom_default_matcher(excluded_headers=headers_to_ignore)
../../../tools/azure-sdk-tools/devtools_testutils/sanitizers.py:103: in set_custom_default_matcher
    _send_matcher_request("CustomDefaultMatcher", {"x-recording-id": x_recording_id}, request_args)
../../../tools/azure-sdk-tools/devtools_testutils/sanitizers.py:563: in _send_matcher_request
    http_client.request(method="POST", url=f"{PROXY_URL}/Admin/SetMatcher", headers=headers_to_send, fields=parameters)
.tox/whl/lib/python3.10/site-packages/urllib3/request.py:78: in request
    return self.request_encode_body(
.tox/whl/lib/python3.10/site-packages/urllib3/request.py:170: in request_encode_body
    return self.urlopen(method, url, **extra_kw)
.tox/whl/lib/python3.10/site-packages/urllib3/poolmanager.py:376: in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
.tox/whl/lib/python3.10/site-packages/urllib3/connectionpool.py:702: in urlopen
    httplib_response = self._make_request(
.tox/whl/lib/python3.10/site-packages/urllib3/connectionpool.py:398: in _make_request
    conn.request(method, url, **httplib_request_kw)
.tox/whl/lib/python3.10/site-packages/urllib3/connection.py:241: in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
/home/pvaneck/.pyenv/versions/3.10.6/lib/python3.10/http/client.py:1282: in request
    self._send_request(method, url, body, headers, encode_chunked)
/home/pvaneck/.pyenv/versions/3.10.6/lib/python3.10/http/client.py:1323: in _send_request
    self.putheader(hdr, value)
.tox/whl/lib/python3.10/site-packages/urllib3/connection.py:226: in putheader
    _HTTPConnection.putheader(self, header, *values)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <urllib3.connection.HTTPSConnection object at 0x7f936ac6aa40>, header = b'x-recording-id', values = [None], i = 0, one_value = None

    def putheader(self, header, *values):
        """Send a request header line to the server.

        For example: h.putheader('Accept', 'text/html')
        """
        if self.__state != _CS_REQ_STARTED:
            raise CannotSendHeader()

        if hasattr(header, 'encode'):
            header = header.encode('ascii')

        if not _is_legal_header_name(header):
            raise ValueError('Invalid header name %r' % (header,))

        values = list(values)
        for i, one_value in enumerate(values):
            if hasattr(one_value, 'encode'):
                values[i] = one_value.encode('latin-1')
            elif isinstance(one_value, int):
                values[i] = str(one_value).encode('ascii')

>           if _is_illegal_header_value(values[i]):
E           TypeError: expected string or bytes-like object
```

After a little bit of debugging, looks like the error stems from a None value for a header being sent [here](https://github.com/Azure/azure-sdk-for-python/blob/9ad85bb9675f08972796fe32fff08eb7b468da2c/tools/azure-sdk-tools/devtools_testutils/sanitizers.py#L561) where the headers are `{'x-abstraction-identifier': 'CustomDefaultMatcher', 'x-recording-id': None}`. Seems like `requests`  drops the `None` header at some point which is why the request works with `requests`, but this `None` value is causing issues when using urllib3 directly. 

Should we ensure that a recording_id is set so that the header is not None (or just not send this header if the recording_id is None)?

File: tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py (hunk 3 above)
Comment (scbedd) [reply tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py:130]: @pvaneck thank you so much for this feedback, I'm going to work off it all and get this thing updated to work properly.

File: tools/azure-sdk-tools/devtools_testutils/sanitizers.py [hunk 4]
```diff
@@ -604,15 +611,14 @@ def _send_sanitizer_request(sanitizer: str, parameters: dict) -> None:
     if is_live_and_not_recording():
         return
 
-    response = requests.post(
+    http_client.request(method="POST", url=
         "{}/Admin/AddSanitizer".format(PROXY_URL),
```
Comment (mccoyp) [inline tools/azure-sdk-tools/devtools_testutils/sanitizers.py:615]: Looks like the URL parameter might have been disrupted by a batch update?
```suggestion
    http_client.request(method="POST",
        url="{}/Admin/AddSanitizer".format(PROXY_URL),
```

Comment (scbedd) [reply tools/azure-sdk-tools/devtools_testutils/sanitizers.py:615]: This looks like a spacing update to me. I'll make certain to apply `black`

File: tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py [hunk 5]
```diff
@@ -109,25 +126,26 @@ def start_record_or_playback(test_id: str) -> "Tuple[str, Dict[str, str]]":
...
```
Comment (mccoyp) [inline tools/azure-sdk-tools/devtools_testutils/proxy_testcase.py:140]: Last instance of using `json=` instead of `body=`

**Comment by scbedd** [comment]
> Hey @scbedd. Ran into some issues with `requests` not being available in certain pipeline environments ([example](https://dev.azure.com/azure-sdk/public/_build/results?buildId=2037071&view=logs&j=4fa73ede-de59-50b0-851d-db96582ea108&t=271dca1a-6c62-5d1b-458d-2dd10be9b8db&l=10994)), and that led me here, which should resolve that issue. However, after running playback tests on a service using this PR (to test it), I ran into the following issue:
> 
> ```
> ../../../tools/azure-sdk-tools/devtools_testutils/proxy_startup.py:193: in start_test_proxy
>     set_custom_default_matcher(excluded_headers=headers_to_ignore)
> ../../../tools/azure-sdk-tools/devtools_testutils/sanitizers.py:103: in set_custom_default_matcher
>     _send_matcher_request("CustomDefaultMatcher", {"x-recording-id": x_recording_id}, request_args)
> ../../../tools/azure-sdk-tools/devtools_testutils/sanitizers.py:563: in _send_matcher_request
>     http_client.request(method="POST", url=f"{PROXY_URL}/Admin/SetMatcher", headers=headers_to_send, fields=parameters)
> .tox/whl/lib/python3.10/site-packages/urllib3/request.py:78: in request
>     return self.request_encode_body(
> .tox/whl/lib/python3.10/site-packages/urllib3/request.py:170: in request_encode_body
>     return self.urlopen(method, url, **extra_kw)
> .tox/whl/lib/python3.10/site-packages/urllib3/poolmanager.py:376: in urlopen
>     response = conn.urlopen(method, u.request_uri, **kw)
> .tox/whl/lib/python3.10/site-packages/urllib3/connectionpool.py:702: in urlopen
>     httplib_response = self._make_request(
> .tox/whl/lib/python3.10/site-packages/urllib3/connectionpool.py:398: in _make_request
>     conn.request(method, url, **httplib_request_kw)
> .tox/whl/lib/python3.10/site-packages/urllib3/connection.py:241: in request
>     super(HTTPConnection, self).request(method, url, body=body, headers=headers)
> /home/pvaneck/.pyenv/versions/3.10.6/lib/python3.10/http/client.py:1282: in request
>     self._send_request(method, url, body, headers, encode_chunked)
> /home/pvaneck/.pyenv/versions/3.10.6/lib/python3.10/http/client.py:1323: in _send_request
>     self.putheader(hdr, value)
> .tox/whl/lib/python3.10/site-packages/urllib3/connection.py:226: in putheader
>     _HTTPConnection.putheader(self, header, *values)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> 
> self = <urllib3.connection.HTTPSConnection object at 0x7f936ac6aa40>, header = b'x-recording-id', values = [None], i = 0, one_value = None
> 
>     def putheader(self, header, *values):
>         """Send a request header line to the server.
> 
>         For example: h.putheader('Accept', 'text/html')
>         """
>         if self.__state != _CS_REQ_STARTED:
>             raise CannotSendHeader()
> 
>         if hasattr(header, 'encode'):
>             header = header.encode('ascii')
> 
>         if not _is_legal_header_name(header):
>             raise ValueError('Invalid header name %r' % (header,))
> 
>         values = list(values)
>         for i, one_value in enumerate(values):
>             if hasattr(one_value, 'encode'):
>                 values[i] = one_value.encode('latin-1')
>             elif isinstance(one_value, int):
>                 values[i] = str(one_value).encode('ascii')
> 
> >           if _is_illegal_header_value(values[i]):
> E           TypeError: expected string or bytes-like object
> ```
> 
> After a little bit of debugging, looks like the error stems from a None value for a header being sent [here](https://github.com/Azure/azure-sdk-for-python/blob/9ad85bb9675f08972796fe32fff08eb7b468da2c/tools/azure-sdk-tools/devtools_testutils/sanitizers.py#L561) where the headers are `{'x-abstraction-identifier': 'CustomDefaultMatcher', 'x-recording-id': None}`. Seems like `requests` drops the `None` header at some point which is why the request works with `requests`, but this `None` value is causing issues when using urllib3 directly.
> 
> Should we ensure that a recording_id is set so that the header is not None (or just not send this header if the recording_id is None)?

Absolutely! Thanks for your write-ups here Paul, effectively all of your suggestions made it into the PR in _some_ fashion.

**Review by pvaneck — APPROVED** [review]
Thanks, @scbedd! Tested locally on some service live/playback tests and everything worked as expected. Assuming the core CI failure is a transient failure, lgtm!

**Pull request merged**